app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  # You should configure your model like this:
+  # devise :omniauthable, omniauth_providers: [:twitter]
+
+  # You should also create an action method in this controller like this:
+  # def twitter
+  # end
+
+  def github
+    callback
+  end
+
+  # More info at:
+  # https://github.com/heartcombo/devise#omniauth
+
+  # GET|POST /resource/auth/twitter
+  # def passthru
+  #   super
+  # end
+
+  # GET|POST /users/auth/twitter/callback
+  # def failure
+  #   super
+  # end
+
+  # protected
+  def callback
+    @user = User.find_or_create_for_oauth(request.env['omniauth.auth'])
+
+    return unless @user.persisted?
```

Comment:
ここで `return` する場合リダイレクトやフラッシュの設定は不要でしょうか？

Reply:
`find_or_create_for_oauth` メソッドで `find_or_create_by` を実行しておりメソッドの結果 `persisted?` がそもそも false になることはありませんでした。
参考にしたサイトの実装の理解が足りていませんでした。

そもそも考慮する必要のない条件分岐であるためこの行自体を削除しました。

Reply:
現状の実装上は確かに保存に失敗するシチュエーションはなさそうですね。ただ、DBに保存する以上保存失敗する可能性はゼロではない(DB側の異常や、今後Userにバリデーション追加して状況が変わる可能性がある)ので、DeviseのWikiを参考に保存失敗時はログインページへリダイレクトすると良いです。

下記、DeviseのWikiから引用です。Facebookの例なので適宜GitHubに置換ください。

```ruby
class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
  # See https://github.com/omniauth/omniauth/wiki/FAQ#rails-session-is-clobbered-after-callback-on-developer-strategy
  skip_before_action :verify_authenticity_token, only: :facebook

  def facebook
    # You need to implement the method below in your model (e.g. app/models/user.rb)
    @user = User.from_omniauth(request.env["omniauth.auth"])

    if @user.persisted?
      sign_in_and_redirect @user, event: :authentication # this will throw if @user is not activated
      set_flash_message(:notice, :success, kind: "Facebook") if is_navigational_format?
    else
      session["devise.facebook_data"] = request.env["omniauth.auth"].except(:extra) # Removing extra as it can overflow some session stores
      redirect_to new_user_registration_url
    end
  end

  def failure
    redirect_to root_path
  end
end
```

Reply:
共有いただいが内容を踏まえた処理に変更しました。
[参考にしたサイト](https://codezine.jp/article/detail/10970)より、Devise の Wiki のコードに寄せた方が一次情報に近しいので望ましいと考え、提出時のコードから修正しました。